### DISCUSSION TITLE
How to automate integration testing 2FA on our hybrid mobile banking App with Google/Duo Authenticator

### DISCUSSION THREAD

**Discussion by idavollen** [appium/appium#19234] [Q&A]
I have a hybrid banding mobile app that supports 2FA by using OTP from Duo/Google Authenticator. I'm wondering how such integration testing between two mobile Apps could be automated with Appium. It'll be much greater if some code example could be found.

**Comment by portyu9** [comment]
Hi @idavollen 

For automated tests, I would not read the OTP from the Google/Duo Authenticator UI.

For a dedicated test account, store its TOTP secret securely in your CI secret store and generate the current RFC-6238 OTP in the test:

```text
test TOTP secret
→ generate current 6-digit OTP
→ enter OTP in banking app
```

This is deterministic and avoids coupling your test to another application's UI.

If you specifically need to test switching between applications, Appium can do that:

```java
driver.activateApp("com.google.android.apps.authenticator2");

// interact if required

driver.activateApp("com.yourcompany.bank");
```

On iOS, use the corresponding bundle IDs.

For Duo Push or other approval flows, use a dedicated test MFA integration/account or backend test hook where available rather than trying to automate a personal authenticator.

So the recommended fix is: **generate TOTP programmatically for integration tests; use `activateApp()` only when cross-app UI behavior itself is what you need to test.**

Please mark this answer as accepted if it helped, thank you.
